File: Examples/intel_examples/examples_core_c/vmlc/source/vzabs.c
/*******************************************************************************
* Copyright 2001-2020 Intel Corporation.
*
* This software and the related documents are Intel copyrighted  materials,  and
* your use of  them is  governed by the  express license  under which  they were
* provided to you (License).  Unless the License provides otherwise, you may not
* use, modify, copy, publish, distribute,  disclose or transmit this software or
* the related documents without Intel's prior written permission.
*
* This software and the related documents  are provided as  is,  with no express
* or implied  warranties,  other  than those  that are  expressly stated  in the
* License.
*******************************************************************************/

/*
!  Content:
!    vzAbs  Example Program Text
!******************************************************************************/

#include <stdio.h>
#include "mkl_vml.h"

#include "_rms.h"

#define INCA 3
#define INCB 5

int main()
{
  MKL_Complex16 zA[10];
  double dB[10];
  double dBha0[10],dBha1[10],dBha2[10];
  double           dBla1[10],dBla2[10];
  double           dBep1[10],dBep2[10];
  float CurRMS,MaxRMS=0.0;

  MKL_Complex16 zA_I[10*INCA];
  double dB_I[10*INCB];
  double dBha0_I[10*INCB],dBha1_I[10*INCB],dBha2_I[10*INCB];
  double                  dBla1_I[10*INCB],dBla2_I[10*INCB];
  double                  dBep1_I[10*INCB],dBep2_I[10*INCB];
  float CurRMS_I,MaxRMS_I=0.0;
  MKL_INT inca=INCA,incb=INCB;

  MKL_INT i=0,vec_len=10;

  zA[0].real=-1000.0000;zA[0].imag=1000.0000;
  zA[1].real=-777.7777;zA[1].imag=777.7777;
  zA[2].real=-555.5555;zA[2].imag=555.5555;
  zA[3].real=-333.3333;zA[3].imag=333.3333;
  zA[4].real=-111.1111;zA[4].imag=111.1111;
  zA[5].real=111.1111;zA[5].imag=-111.1111;
  zA[6].real=333.3333;zA[6].imag=-333.3333;
  zA[7].real=555.5555;zA[7].imag=-555.5555;
  zA[8].real=777.7777;zA[8].imag=-777.7777;
  zA[9].real=1000.0000;zA[9].imag=-1000.0000;
  dB[0]=1.4142135623730951e+003;
  dB[1]=1.0999437718513525e+003;
  dB[2]=7.8567412275096603e+002;
  dB[3]=4.7140447365057963e+002;
  dB[4]=1.5713482455019320e+002;
  dB[5]=1.5713482455019320e+002;
  dB[6]=4.7140447365057963e+002;
  dB[7]=7.8567412275096603e+002;
  dB[8]=1.0999437718513525e+003;
  dB[9]=1.4142135623730951e+003;

  for(i=0;i<10;i++) {
    zA_I[i*inca]=zA[i];
    dB_I[i*incb]=dB[i];
  }

  vzAbs(vec_len,zA,dBha0);
  vzAbsI(vec_len,zA_I,inca,dBha0_I,incb);

  vmzAbs(vec_len,zA,dBep1,VML_EP);
  vmzAbsI(vec_len,zA_I,inca,dBep1_I,incb,VML_EP);

  vmlSetMode(VML_EP);
  vzAbs(vec_len,zA,dBep2);
  vzAbsI(vec_len,zA_I,inca,dBep2_I,incb);

  vmzAbs(vec_len,zA,dBla1,VML_LA);
  vmzAbsI(vec_len,zA_I,inca,dBla1_I,incb,VML_LA);

  vmlSetMode(VML_LA);
  vzAbs(vec_len,zA,dBla2);
  vzAbsI(vec_len,zA_I,inca,dBla2_I,incb);

  vmzAbs(vec_len,zA,dBha1,VML_HA);
  vmzAbsI(vec_len,zA_I,inca,dBha1_I,incb,VML_HA);

  vmlSetMode(VML_HA);
  vzAbs(vec_len,zA,dBha2);
  vzAbsI(vec_len,zA_I,inca,dBha2_I,incb);

  for(i=0;i<10;i++) {
    if(dBha0[i]!=dBha1[i] || dBha1[i]!=dBha2[i]) {
      printf("Error! Difference between vzAbs and vmzAbs in VML_HA mode detected.\n");
      return 1;
    }

    if(dBha0_I[i*incb]!=dBha1_I[i*incb] || dBha1_I[i*incb]!=dBha2_I[i*incb]) {
      printf("Error! Difference between vzAbsI and vmzAbsI in VML_HA mode detected.\n");
      return 1;
    }

    if(dBla1[i]!=dBla2[i]) {
      printf("Error! Difference between vzAbs and vmzAbs in VML_LA mode detected.\n");
      return 1;
    }

    if(dBla1_I[i*incb]!=dBla2_I[i*incb]) {
      printf("Error! Difference between vzAbsI and vmzAbsI in VML_LA mode detected.\n");
      return 1;
    }

    if(dBep1[i]!=dBep2[i]) {
      printf("Error! Difference between vzAbs and vmzAbs in VML_EP mode detected.\n");
      return 1;
    }

    if(dBep1_I[i*incb]!=dBep2_I[i*incb]) {
      printf("Error! Difference between vzAbsI and vmzAbsI in VML_EP mode detected.\n");
      return 1;
    }
  }

  printf("vzAbs test/example program\n\n");
  printf("           Argument                           vzAbs\n");
  printf("===============================================================================\n");
  for(i=0;i<10;i++) {
    printf("   % .4f %+.4f*i      % .10f\n",zA[i].real,zA[i].imag,dBha0[i]);
    CurRMS=drelerr(dB[i],dBha0[i]);
    if(CurRMS>MaxRMS) MaxRMS=CurRMS;
  }
  printf("\n");
  printf("vzAbsI test/example program\n\n");
  printf("           Argument                           vzAbs\n");
  printf("===============================================================================\n");
  for(i=0;i<10;i++) {
    printf("   % .4f %+.4f*i      % .10f\n",zA_I[i*inca].real,zA_I[i*inca].imag,dBha0_I[i*incb]);
    CurRMS_I=drelerr(dB_I[i*incb],dBha0_I[i*incb]);
    if(CurRMS_I>MaxRMS_I) MaxRMS_I=CurRMS_I;
  }
  printf("\n");
  if(MaxRMS>=1e-5) {
    printf("Error! Relative accuracy is %.16f\n",MaxRMS);
    return 1;
  }
  else {
    printf("Relative accuracy is %.16f\n",MaxRMS);
  }

  printf("\n");

  if(MaxRMS_I>=1e-5) {
    printf("Error! Relative strided accuracy is %.16f\n",MaxRMS_I);
    return 1;
  }
  else {
    printf("Relative strided accuracy is %.16f\n",MaxRMS_I);
  }
  return 0;
}

File: Examples/intel_examples/examples_core_f/vmlf/source/vzabs.f
!===============================================================================
! Copyright 2001-2020 Intel Corporation.
!
! This software and the related documents are Intel copyrighted  materials,  and
! your use of  them is  governed by the  express license  under which  they were
! provided to you (License).  Unless the License provides otherwise, you may not
! use, modify, copy, publish, distribute,  disclose or transmit this software or
! the related documents without Intel's prior written permission.
!
! This software and the related documents  are provided as  is,  with no express
! or implied  warranties,  other  than those  that are  expressly stated  in the
! License.
!===============================================================================

!  Content:
!    vzAbs  Example Program Text
!*******************************************************************************

      include "_rms.fi"

      program MKL_VML_TEST

      include "mkl_vml.f90"
      real(kind=4) :: srelerr
      real(kind=8) :: drelerr
      real(kind=4) :: crelerr
      real(kind=8) :: zrelerr

      complex(kind=8) zA(10)
      real(kind=8) dB(10)
      real(kind=8) dBha0(10)
      real(kind=8) dBha1(10)
      real(kind=8) dBha2(10)
      real(kind=8) dBla1(10)
      real(kind=8) dBla2(10)
      real(kind=8) dBep1(10)
      real(kind=8) dBep2(10)
      real(kind=8) CurRMS,MaxRMS

      integer, parameter :: inca=3
      integer, parameter :: incb=5
      complex(kind=8) zA_I(10*inca)
      real(kind=8) dB_I(10*incb)
      real(kind=8) dBha0_I(10*incb)
      real(kind=8) dBha1_I(10*incb)
      real(kind=8) dBha2_I(10*incb)
      real(kind=8) dBla1_I(10*incb)
      real(kind=8) dBla2_I(10*incb)
      real(kind=8) dBep1_I(10*incb)
      real(kind=8) dBep2_I(10*incb)
      real(kind=8) CurRMS_I,MaxRMS_I

      integer(kind=8) mode
      integer tmode
      integer i, vec_len

      vec_len=10
      MaxRMS=0.0
      MaxRMS_I=0.0

      zA( 1)=(-1000.0000d+000,1000.0000d+000)
      zA( 2)=(-777.7777d+000,777.7777d+000)
      zA( 3)=(-555.5555d+000,555.5555d+000)
      zA( 4)=(-333.3333d+000,333.3333d+000)
      zA( 5)=(-111.1111d+000,111.1111d+000)
      zA( 6)=(111.1111d+000,-111.1111d+000)
      zA( 7)=(333.3333d+000,-333.3333d+000)
      zA( 8)=(555.5555d+000,-555.5555d+000)
      zA( 9)=(777.7777d+000,-777.7777d+000)
      zA(10)=(1000.0000d+000,-1000.0000)
      dB( 1)=1.4142135623730951d+003
      dB( 2)=1.0999437718513525d+003
      dB( 3)=7.8567412275096603d+002
      dB( 4)=4.7140447365057963d+002
      dB( 5)=1.5713482455019320d+002
      dB( 6)=1.5713482455019320d+002
      dB( 7)=4.7140447365057963d+002
      dB( 8)=7.8567412275096603d+002
      dB( 9)=1.0999437718513525d+003
      dB(10)=1.4142135623730951d+003

      do i=1,10
          zA_I((i-1)*inca+1)=zA(i)
          dB_I((i-1)*incb+1)=dB(i)
      end do

      call VZABS(vec_len,zA,dBha0)
      call VZABSI(vec_len,zA_I,inca,dBha0_I,incb)

      mode=VML_EP
      call VMZABS(vec_len,zA,dBep1,mode)
      call VMZABSI(vec_len,zA_I,inca,dBep1_I,incb,mode)

      tmode=VML_EP
      tmode=VMLSETMODE(tmode)
      call VZABS(vec_len,zA,dBep2)
      call VZABSI(vec_len,zA_I,inca,dBep2_I,incb)

      mode=VML_LA
      call VMZABS(vec_len,zA,dBla1,mode)
      call VMZABSI(vec_len,zA_I,inca,dBla1_I,incb,mode)

      tmode=VML_LA
      tmode=VMLSETMODE(tmode)
      call VZABS(vec_len,zA,dBla2)
      call VZABSI(vec_len,zA_I,inca,dBla2_I,incb)

      mode=VML_HA
      call VMZABS(vec_len,zA,dBha1,mode)
      call VMZABSI(vec_len,zA_I,inca,dBha1_I,incb,mode)

      tmode=VML_HA
      tmode=VMLSETMODE(tmode)
      call VZABS(vec_len,zA,dBha2)
      call VZABSI(vec_len,zA_I,inca,dBha2_I,incb)

      do i=1,10
          if(dBha0(i) .ne. dBha1(i)) then
              print *,"Error! Difference between VZABS and"
              print *," VMZABS in VML_HA mode detected"
              stop 1
          endif
          if(dBha0_I((i-1)*incb+1) .ne. dBha1_I((i-1)*incb+1)) then
              print *,"Error! Difference between VZABSI and"
              print *," VMZABSI in VML_HA mode detected"
              stop 1
          endif
          if(dBha1(i) .ne. dBha2(i)) then
              print *,"Error! Difference between VZABS and"
              print *," VMZABS in VML_HA mode detected"
              stop 1
          endif
          if(dBha1_I((i-1)*incb+1) .ne. dBha2_I((i-1)*incb+1)) then
              print *,"Error! Difference between VZABSI and"
              print *," VMZABSI in VML_HA mode detected"
              stop 1
          endif
          if(dBla1(i) .ne. dBla2(i)) then
              print *,"Error! Difference between VZABS and"
              print *," VMZABS in VML_LA mode detected"
              stop 1
          endif
          if(dBla1_I((i-1)*incb+1) .ne. dBla2_I((i-1)*incb+1)) then
              print *,"Error! Difference between VZABSI and"
              print *," VMZABSI in VML_LA mode detected"
              stop 1
          endif
          if(dBep1(i) .ne. dBep2(i)) then
              print *,"Error! Difference between VZABS and"
              print *," VMZABS in VML_EP mode detected"
              stop 1
          endif
          if(dBep1_I((i-1)*incb+1) .ne. dBep2_I((i-1)*incb+1)) then
              print *,"Error! Difference between VZABSI and"
              print *," VMZABSI in VML_EP mode detected"
              stop 1
          endif
      end do

      print *,"vzAbs test/example program"
      print *,""
      print *,"           Argument                     vzAbs"
      print *,"======================================================", &
     &        "========================"
      do i=1,vec_len
            print 10,zA(i),"    ",dBha0(i)
            CurRMS=drelerr(dB(i),dBha0(i))
            if(CurRMS>MaxRMS) MaxRMS=CurRMS
      end do

      print *,""
      print *,"vzAbsI test/example program"
      print *,"           Argument                     vzAbs"
      print *,"======================================================", &
     &        "========================"
      do i=1,10
            print 10,zA_I((i-1)*inca+1),"    ",dBha0_I((i-1)*incb+1)
            CurRMS_I=drelerr(dB_I((i-1)*incb+1),dBha0_I((i-1)*incb+1))
            if(CurRMS_I>MaxRMS_I) MaxRMS_I=CurRMS_I
      end do

      print *,""
      if(MaxRMS>=1e-14) then
            print 11,"Error! Relative accuracy is ",MaxRMS
            stop 1
      else
            print 11,"Relative accuracy is ",MaxRMS
      endif

      if(MaxRMS_I>=1e-14) then
            print 11,"Error! Relative strided accuracy is ",MaxRMS_I
            stop 1
      else
            print 11,"Relative strided accuracy is ",MaxRMS_I
      endif

10    format(E15.7,E15.7,A5,E15.7)
11    format(A,F25.16)

      end
